API

sd-tomita opened the pull request:
## TODO
- [x] Sdx.Api.Json 用のユニットテスト試作
- [x] Sdx.Api.Json は Sdx.Web の中に移動
- [x] Sdx.Web.Url の作成(URLの解析・組み立て)
  - [x] 変数をキャメルケースにする(スネークケースは使わない)
  - [x] Get、Set メソッド廃止してプロパティを使うようにする
  - [x] ~~Sdx.Web.Url.Protocol を追加する(列挙型を使う)~~ `System.Uri.Scheme` で取れるのでなし
  - [x] ~~Build() の引数には可変長引数を使って色々受け取れるようにする~~ 使わないことになったのでなし
  - [x] StringBuilder による文字列連結
  - [x] foreach → ForEach / LINQ に書き換え(LINQはできるかどうかの確認も含め)
  - [x] テストケースの追加
  - [x] 追加したテストに伴うコードの修正(もしあれば)
- [x] Sdx.Web は Sdx プロジェクト内へ移動する

以下は別projで対応する
- [ ] Sdx.Web.Api の作成(APIへのリクエスト送信、受信データを返す。全ての機能を備える必要は無いが、HttpClient のインターフェースを参考に作成する。メソッド名とか同じでもOK)

sd-tomita committed:
ポイントAPIへのリクエスト管理ライブラリ用プロジェクトを新規作成

sd-tomita committed:
APIテスト用のクラスを追加

sd-tomita committed:
テスト用のテンプレートをまずはそのままコピペ

sd-tomita commented:
## TODO

テストをまず先に作って、一旦見てもらう。

sd-tomita committed:
クラス名が雛形のままだったのでとりあえず変更

sd-tomita commented:
## メモ
### 前提
事前に一度ユニットテストの debug ビルドが必要です。debug じゃないとテストできません。

### テストの実行方法

[テスト]-[ウィンドウ]-[テストエクスプローラー] でテストエクスプローラーを有効にし、 エクスプローラー上のいづれかの実行方法をクリック(すべて実行、未実行だけ実行、など)

便利なショートカットキー
https://msdn.microsoft.com/ja-jp/library/da5kh0wa.aspx#bkmk_testexplorerGLOBAL

### `[Fact]` などの表記について

属性の自作
http://ufcpp.net/study/csharp/sp_attribute.html#userdefine
cs-sdx ではテスト実行時に`[Fact]` に書いた動作がテストされる。

sd-tomita committed:
Fact 以外は一旦削除

sd-tomita committed:
テスト試し書き

とりあえずこうやって使えたらいいなという内容で
書いてみました。

gomo commented on UnitTest/ApiUseTest.cs:
```diff
@@ -0,0 +1,46 @@
+        ;
+
+      //結果配列を取得する
+      var Resp = Api->GetResponse();
```
``` c#
var api = new Sdx.Api.Json("http://example.com/path/to/api");

api.Post
  .Set("client_id", "1")
  .Set("shop_id", "1234")
  .Set("access_token", "asdfghjkl");

api.Get
  .Set("id", "1")

```

sd-tomita commented:
## メモ

汎用的にするにあたっていくつか課題が判明したのでメモ
- 色んなドメインでも使えるようにする
- 特定のAPIだけに使うわけではないので、SetXXXX によるメソッド名の決め打ちはしないほうがよさげ
- POST or GET ? このあたりをどうコントロールできるようにするか
- GetResponse() だとHttpリクエスト自体のレスポンスを取得できそうなイメージなので、実際の動きとの乖離が生じるので違う名前なりなんなりしたほうがよい
- そもそもリクエストやレスポンスの組み立ては ASP.net の既存クラスでできるんじゃないか？←いまここ

sd-tomita committed:
文法間違いだけ取り急ぎ修正

PHPの書き方になってた

sd-tomita commented:
## リクエストを投げる

.NET Framework 4.5 以上ではあるものの、こんな方法があるみたいです。
[参考サイト](http://blog.shibayan.jp/entry/20120822/1345563275)

``` C#
var params = new FormUrlEncodedContent(new Dictionary<string, string>
{
    {"key1", param1},
    {"key2", param2},
    {"key3", param3}
});
HttpClient client = new HttpClient();
HttpResponseMessage response= await client.PostAsync(new Uri("http://example.com/path/to/api"), params);
response.Content.ReadAsStringAsync();//=>api の取得結果
```

※追記

```
警告  
参照されているプロジェクト 'Sdx.Api' は、より新しいバージョン (4.5) のフレームワークをターゲットと
しています。このプロジェクトの現在のターゲット フレームワークのバージョンは 4.0 です。
このプロジェクトのターゲット フレームワークの範囲外のアセンブリの種類が依存関係チェーンの
いずれかのプロジェクトで使用されている場合、バージョンの違いが原因でビルドに失敗する可能性が
あります。
```

sd-tomita committed:
UnitTest プロジェクトに Sdx.Api への参照を追加

Visual Studio でずっと参照不足が表示されていた

sd-tomita commented:
> 現在のターゲット フレームワークのバージョンは 4.0 です。

確認したらやっぱり 4.5 はだめでした。残念。

sd-tomita commented:
## 気をとりなおして

4.0 でも動く方法を探してみました。
まだ動かしていませんが、コードだけメモします。
レスポンス操作についてはまた別途。

``` C#
using System.Net;

//リクエスト用のインスタンス生成
HttpWebRequest request = WebRequest.Create("http://example.com/path/to/api");

//パラメータ
string params = "key1=aaa&key2=bbb&key3=ccc";

//設定変更
request.Method = "POST";
request.ContentType = "application/x-www-form-encoded";
var stream = request.GetRequestStream();

//パラメータを付与する
var writer = new StreamWriter(stream, Encoding.UTF8);
writer.WriteLine(params);

//レスポンス取得
WebResponse response = request.GetResponse();
```

参考：
http://www.atmarkit.co.jp/fdotnet/dotnettips/318webpost/webpost.html
ネットで検索した限りでは、見つかった記事はどこもわりと似たような手法だったので、これが定石のようなものなのかなと思います。

sd-tomita commented:
## サンプル

JSON 形式の固定値を単純に返すだけのページを使って、コンソールアプリでテスト。
このへんが基本動作になる予定です。

``` C#
using System;
using System.IO;
using System.Text;
using System.Net;

namespace ApiTest
{
  public class Program
  {
    public static void Main()
    {
      //リクエスト用のインスタンス生成
      WebRequest request = WebRequest.Create("http://local.www.site.com/");

      //設定変更
      request.Method = "POST";
      request.ContentType = "application/x-www-form-encoded";
      Stream stream = request.GetRequestStream();

      //パラメータを付与する(※今回は特にパラメータは使っていませんが付与する場合の書き方)
      string param = "key1=aaa&key2=bbb&key3=ccc";
      StreamWriter writer = new StreamWriter(stream, Encoding.UTF8);
      writer.WriteLine(param);

      //レスポンス取得
      WebResponse response = request.GetResponse();
      Stream stream2 = response.GetResponseStream();

      //レスポンスデータを読むためのStreamReaderを作成
      StreamReader reader = new System.IO.StreamReader(stream2, Encoding.UTF8);

      //データをすべて読みこむ
      string result = reader.ReadToEnd();

      //閉じる
      response.Close();

      //出力
      Console.WriteLine(result);
    }
  }
}
```

結果

```
{ "name": "わたくし", "age": 33, "married": false }
```

sd-tomita commented:
タスク: Sdx.Api.Json は Sdx.Web の中に移動します。

sd-tomita committed:
APIクラスのテスト内容を修正しました。

リクエスト先と POST/GET の指定を外から渡せるようにしました

sd-tomita commented:
`response.Close()` は呼ばないで、using を使う
https://github.com/SunriseDigital/CsharpSunrise/issues/1#issuecomment-91105558

sd-tomita committed:
Api用のクラスは Sdx.Web の中へ移動する

使わなくなったクラスは削除予定です。
(然るべき手順で削除しないと不要なファイル等が残ったり、cs-sdx.sln などが
変な状態になりそうなので、ちゃんと調べてから対応します)

sd-tomita committed:
テストコード追記

APIリクエストを投げるだけのメソッドと
通信の成否を問うメソッドもあったほうがいいかも　と思い追加。
if 文で使ったりできそう。

gomo commented on UnitTest/ApiUseTest.cs:
```diff
@@ -0,0 +1,52 @@
+    {
+      Console.WriteLine("TestMethod1");
+
+      Sdx.Api.Json api = new Sdx.Api.Json("http://example.com/path/to/api");
```
``` c#
var url = new Sdx.Web.Url("http://example.com/path/to/api");
var api = new Sdx.Api.Json(url);
```

gomo replied:
``` c#
var url = new Sdx.Web.Url("http://example.com/path/to/api?foo=bar&hoge=huga");

Assert.Equal("example.com", url.GetDomain());
Assert.Equal(Sdx.Web.Url.Protocol.Http, url.GetProtocol());
Assert.Equal("bar", url.GetParam("foo"));
Assert.Equal("huga", url.GetParam("hoge"));
Assert.Equal("/path/to/api", url.GetPath());

url.SetParam("key", "value");


Assert.Equal("http://example.com/path/to/api?foo=bar&hoge=huga&key=value", url.Build());

var param = new Dictionary<string, string>();
param["new"] = "newValue";
Assert.Equal("http://example.com/path/to/api?foo=bar&hoge=huga&key=value&new=newValue", url.Build(param));
Assert.Equal("http://example.com/path/to/api?foo=bar&hoge=huga&key=value", url.Build());

var array = new String[]{"key"};
Assert.Equal("http://example.com/path/to/api?foo=bar&hoge=huga", url.Build(array));
Assert.Equal("http://example.com/path/to/api?foo=bar&hoge=huga&key=value", url.Build());
```

gomo commented on UnitTest/ApiUseTest.cs:
```diff
@@ -0,0 +1,52 @@
+      ;
+
+      //リクエストを投げる
+      api.Run();
```
ここまではJson関係ないので、HttpClient的な何かにしましょう。

で

``` c#
var resp = api.Send();
```

みたいな感じでレスポンスオブジェクトを受け取る感じで。

gomo commented on UnitTest/ApiUseTest.cs:
```diff
@@ -0,0 +1,52 @@
+      {
+        //結果配列を取得する
+        var json = api.GetJson();
+      }
```
``` c#
if(resp.IsSuccess())
{
  var input = new StringReader(api.GetBody());
  Sdx.Config.Tree json = new Sdx.Config.TreeJson(input);
}
```

Sdx.Config.Treeが使えるかどうかわからないけど、行けそうな気がするんだけど・・・

sd-tomita replied:
このあたりは受信データをどうするか の段階まで来たらまた相談しますー

sd-tomita committed:
Sdx.Api を削除しました

bece3b1 で Sdx.Web に移動しているため
こちらは削除します

sd-tomita commented:
## ソリューション項目の削除

https://msdn.microsoft.com/ja-jp/library/vstudio/1ee8zw5t(v=VS.100).aspx
関連記事抜粋

> [ソリューション項目] フォルダーから項目を削除できます。 項目の削除とは、項目を完全に削除することではありません。 項目の削除では、項目とソリューションとの関連付けが解除されます。 項目を表すファイルは、ディスクに残ります。
> 1. ソリューション項目を削除するにはソリューション エクスプローラーで、削除する項目を選択します。
> 2. [編集] メニューの [削除] をクリックします。
> 3. 項目に対応するファイルが残っている限り、その項目を再度追加できます。
> 
> ヒント
> ストレージからファイルを完全に削除するには、Windows エクスプローラーを使用します。

sd-tomita committed:
Sdx.Url を追加

ついでにユニットテスト用のクラスと、UnitTestに
Sdx.Urlへの参照設定も追加

sd-tomita committed:
Sdx.Url の作る場所間違えていたので一旦削除

Sdx.Web 内に作るんだった。

sd-tomita committed:
Sdx.Web.Url クラスを追加

作るところ間違えていて一旦削除したので改めて作成

gomo commented:
> cs-sdx ではテスト実行時に[Fact] に書いた動作がテストされる。

厳密に言うとxunitの機能です。
https://github.com/xunit/xunit

travisのテストがマイクロソフト純正のMSTestに対応してなかったので、一番新しいテストユニットのxunitを採用したのですが、今度はxunitがVisualStudioとの相性が悪くて、苦労しました。

テストのテンプレートクラスにごちゃごちゃ書いてあるのはそういうわけです。

sd-tomita committed:
Sdx.Url のテスト用クラスを追加

アセンブリ足りない分はおいおい追加する

sd-tomita committed:
System.Collections.Generic の参照を追加

sd-tomita committed:
コメント内容を一部補足

sd-tomita committed:
API クラスのテストが始まるまでは一旦コメントアウトします

ユニットテストでエラーが出続けてテストにならないので、
Sdx.Web.Url の作業が終わって Sdx.Web.Api の作業に入るまでは
Url側のテストのためとりあえずコメントアウトします。

sd-tomita committed:
クラスにアクセス修飾子が無くてコンストラクターが見つからなかったのを修正

クラスにアクセス修飾子を付ける習慣がなかった。反省。

sd-tomita committed:
一旦全てのテストをコメントアウト

まずはコンストラクタ動作のみ確認するため。
実装ごとにコメントアウトをひとつづつ外していきます

sd-tomita committed:
UrlTest分のソリューションファイル更新

sd-tomita committed:
コンストラクタの内部動作を追加

直接 System.Uri クラス使ってもよかったかもしれませんが
クエリ文字列からキーごとに値を取得するのは System.Uri クラスの
メソッドではできなさそう？だったので、とりあえずSdx.Web.Uriのコンストラクターで
インスタンスの生成をして、パラメータは自前で連想配列を作成してキーごとに値を格納して
みました。

sd-tomita committed:
ドメイン取得用メソッドを追加

sd-tomita committed:
名前が微妙だったので変更

これでも微妙っちゃ微妙だけど、query って名前だと
"?key=value" がそのまま入っているような気がしたので
名前を変えてみましたｓ

sd-tomita committed:
パラメータの個別取得用メソッドを追加

sd-tomita committed:
テストでキーが見つからないエラーが出ていたのを修正

`?` が入ったままだったのでカット。
System.Uri.Query から取り出してそのまま使っているので
そりゃ残っているか。

sd-tomita committed:
GetParam テスト用のコメントアウトを解除

テストも通ったことを確認

sd-tomita committed:
DocumentRoot 以下のパスを取得するメソッドを追加

sd-tomita committed:
通信プロトコル名を取得するメソッドを追加

Sdx.Web.Url.Protocol.Http のところの書き方がよくわからないので
とりあえずテストは後回し

sd-tomita committed:
パラメータのセッターを追加

sd-tomita committed:
URLの組み立て用メソッドを追加

sd-tomita committed:
生成されたクエリ文字列がおかしかったので修正

sd-tomita committed:
URLが間違っていたので修正

ひどいことになってたｗ

sd-tomita committed:
SetParam, Build のテストを有効にしました

sd-tomita committed:
public にする必要なかったのでアクセス修飾子変更

sd-tomita committed:
build() の引数に連想配列を渡した際の挙動を追加

メソッド内で引数でうまく分岐する書き方がわからなかったので
とりあえずオーバーロードしています。

sd-tomita committed:
path の生成動作をメソッドに移行

同じコードなので。

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -0,0 +1,96 @@
+    public Url(string url_str)
+    {
+      //@System.Uri
+      this._uri = new Uri(url_str);
```
private/protectedに_は付けない方針でやっています。マイクロソフトの方針に従っています。
:innocent:

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -0,0 +1,96 @@
+      foreach (var item in param_list)
+      {
+        string[] item_part = item.Split('=');
+        data.Add(item_part[0], item_part[1]);
```
Addは同じキーが既に存在した時に例外になるのでこの場合はふさわしくないと思います。SetParamもですね。
:innocent:

sd-tomita replied:
※口頭で聞いた話のメモ
インデクサ(`data[key]`)に修正します。

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -0,0 +1,96 @@
+      {
+        query += string.Format("{0}={1}&", pair.Key, pair.Value);
+      }
+      return query;
```
これだと最後に&がついちゃうのを、buildでtrimしてるようですが、このメソッドの責任的にここでtrimするべきで後思います。

:innocent:

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -0,0 +1,96 @@
+      this._param_data.Add(key, value);
+    }
+
+    private string _BuildQueryString(Dictionary<string,string> param)
```
URLクエリーは、追加するのと、除くのと、2つの機能がありますが、その機能はクエリーを組み立てる部分に持たせるのがいいと思います。

除く機能はWhereとかSelectとかといったLINQ系メソッドを追加うと便利です。
http://qiita.com/RyotaMurohoshi/items/118d4aa4d5be8c71c9ce

:innocent:

sd-tomita committed:
同じ key を指定した場合にエラーにならないようにするための変更

Dictionary.Add() ではなくインデクサを使うようにしました。

sd-tomita committed:
private/protected メソッド、メンバ変数の名前を修正

_ は使わないようにしました

sd-tomita committed:
Build() と BuildQueryString() の内容を少し整理

クエリ文字列の生成は BuildQueryString内で完結できるように。

sd-tomita commented:
パラメータの順番が変わってテストがエラーになっていた。修正する。(Travisが吐いているエラーもそれ。)

```
 Expected: http://example.com/path/to/api?foo=bar&hoge=huga&key=value&new=newValue
 Actual:   http://example.com/path/to/api?new=newValue&foo=bar&hoge=huga&key=value
```

sd-tomita committed:
パラメータの指定順を維持するための変更

sd-tomita commented:
### TODO

以下の2点進める。うまくいきそうでいかない。
- [x] `BuildQueryString()` メソッドの引数を object 型にして、内部で型を見て処理を分岐する
- [x] List<string> 型の引数を `Build()` メソッドに渡したとき、LINQのメソッドを使ってクエリから取り除く

sd-tomita committed:
色々途中ですが、リモートに記録を残してはおきたいので一旦コミット

sd-tomita commented on Sdx.Web/Url.cs:
```diff
@@ -0,0 +1,138 @@
+      if(param is Dictionary<string, string>)
+      {
+        //キャストしないとコンパイルエラーになる。。。微妙なので違う方法を探す
+        var superparam = (Dictionary<string, string>)param;
```
コンパイル時は型が確定していないから、`object param` を `foreach` で回そうとしてエラーになる。
という当たり前っぽい事実に今頃気づく。 object 型の持つメソッド等(例えば `.ToString()` とか)でなんとかできるような動作以外は基本的にキャストするか、そもそも Object型は使わない とするしかなさそう。

gomo replied:
強く片づけされた言語では、変数に型があって、変数に代入するとその方のメソッドしか呼び出せなくなります。

実はオブジェクト指向ではこっちのほうが当たり前で、PHPでクラスを作るときも、同じような考えで設計する必要があります。

この考え方って、先日ミーティングで話しがでたオブジェクト指向の原則に則った考え方なんですけど、どれだかわかりますか？

gomo replied:
> キャストするか、そもそも Object型は使わない とするしかなさそう。

オブジェクト型を使う時はオブジェクト型である意味があるときです。IF文で型をチェックしてキャストするコードが現れた時は、なにかオブジェクト指向の原則に反することをしている黄色信号です。基本的に公開インターフェースについてはオーバーロードで型を限定するか、ジェネリック型のメソッドやクラスをつかいます。

今回はprivateなメソッドなので、オーバーロードでいくつもメソッド書くより、読みやすくなるから、そうしただけです。

sd-tomita committed:
キャストの理由を説明したコメントを修正

原因がわかったため、コメントにもその旨追記しました。
あとついでに変数名も修正しました。

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -0,0 +1,138 @@
+        {
+          param_list[item.Key] = item.Value;
+        }
+      }
```
計算量を考えると

`this.param_data`がからの時`O(1)`、`this.param_data`がからじゃない時`O(1) + O(n)`

このIF分を外すと

`this.param_data`がからの時`O(1)`、`this.param_data`がからじゃない時`O(n)`

です。どちらがいいとお思いますか？

gomo replied:
計算量は変わらないけどコピーコンストラクタを使ったほうがコードが簡潔だと思われる。

``` c#
var param_list = new Dictionary<string, string>(this.param_data);
```

ちなみにシャローコピーなので気をつけてください。

gomo commented:
LINQを使って辞書をフィルタリングするコード。

``` c#
var param_list = this.param_data
  .Where(kv => exclude_list.Contains(kv.Key))
  .ToDictionary(kv => kv.Key, kv => kv.Value);
```

sd-tomita committed:
クエリ文字列生成処理を全てメソッドにまとめました

Build() からは引数変えて BuildQueryString() を呼ぶだけで済むようになりました

sd-tomita committed:
変数の名前が微妙だったので修正

このあと引数が Dictionary 型じゃなくて List 型のバージョンも
書くので、その時に使いそうな listParams という名前はやめました

gomo commented on UnitTest/UrlTest.cs:
```diff
@@ -0,0 +1,58 @@
+      //Assert.Equal(Sdx.Web.Url.Protocol.Http, url.GetProtocol());
+      Assert.Equal("bar", url.GetParam("foo"));
+      Assert.Equal("huga", url.GetParam("hoge"));
+      Assert.Equal("/path/to/api", url.GetPath());
```
`url.Domain`
`url.Parameters["foo"]`
`url.Path`

sd-tomita committed:
Build() の引数に List<string> 型のデータを渡した際の挙動を追加

同じような処理を書いている箇所がいくつかあるので、まとめる余地はありそうですが
一旦コミットします。

テストを勘違いしていて、テストでは
List<string> ではなく string[] 型のデータを渡していたので
現段階ではテストがコケます。

sd-tomita commented:
### 追加TODO

※最上段コメント内に移動します。

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -0,0 +1,129 @@
+  public class Url
+  {
+    private Uri uri;
+    private Dictionary<string, string> param_data;
```
細かいですけど、メンバー変数とローカル変数はキャメルケースで書くようにしています。マイクロソフトの推奨だったので。

まあ、公開されるインターフェースじゃないので余り神経質にしなくてもいいけど。

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -0,0 +1,129 @@
+    private Dictionary<string, string> param_data;
+
+    //コンストラクタ
+    public Url(string url_str)
```
あ、publicメソッドの引数名は公開メンバーになります。名前を指定して引数を指定する機能があって、この変数名が名前になるので、これを変更すると、使っている場所は死亡します。

この前安西さんと相談して、ひとまず名前付き引数呼び出しはしない方向がいいのでは？という話になっていますが、非常に引数が多く、省略できる引数が幾つもある場合、便利なこともあるので、呼ばれることは想定しておいたほうがいいと思います。

引数はキャメルケース統一でお願いします。

:innocent:

sd-tomita replied:
了解です。 [このコメント](https://github.com/SunriseDigital/cs-sdx/pull/16#discussion_r39115500) の件もついでに合わせて修正します。

名前についてはこのあたりがマイクロソフトの推奨ですかね。
https://msdn.microsoft.com/ja-jp/library/ms229043.aspx

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -0,0 +1,129 @@
+    private string BuildQueryString(object param = null)
+    {
+      //ここに格納された値を最終的に "&" で連結してクエリ文字列にします
+      var baseParams = new List<string>();
```
Listを一回作る意味が余り無さそうな気がしました。

``` c#
string.Join("&", baseParams);
```

というのは一行ですが、多分計算量はO(n)です。

いきなり文字列で連結したほうが効率いいと思います。一番簡単な文字列の連結は

``` c#
var query = "?";

query += "foo=bar";
```

ですが、今回のように繰り返しやる場合、StringBuilderを使ったほうがパフォーマンスがいいらしいですね。調べて、それを使ってください。

:innocent:

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -0,0 +1,129 @@
+    {
+      //ここに格納された値を最終的に "&" で連結してクエリ文字列にします
+      var baseParams = new List<string>();
+      foreach (KeyValuePair<string, string> item in this.param_data)
```
c#のforeachはwhile文に展開されます。

``` c#
IEnumerator e = array.GetEnumerator();
while(e.MoveNext())
{
  型名 変数 = (型名)e.Current;
  文
}
```

CurrentやMoveNextのオーバヘッドが高いらしく余りパフォーマンスが良くないらしいので、ForEachメソッドがつかえたらソッチのほうがいいみたいですね。
http://ufcpp.net/study/csharp/sp_foreach.html

しかもForEachはLINQに置き換えられる時が多く、そのほうが完結で観やすかったりするらしいです。
http://qiita.com/t_takahari/items/6dc72f48b1ebdfed93b7

使えるかどうか試してください。

:innocent:

sd-tomita replied:
了解です。LINQ に置き換えられるかは確認して、できそうなところは書き換えてみます。

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -0,0 +1,129 @@
+        }
+      }
+
+      if(param is List<string>)
```
ここは`else if`でOKですね。というかeles ifじゃないとダメですね。余計な計算が発生します。

sd-tomita replied:
たしかに。。。
別件の修正作業で BuildQueryString() 内から if 文はなくなるので、ここはひとまずそのままにします。

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -0,0 +1,129 @@
+      string path = this.BuildPath();
+      string query = this.BuildQueryString(param);
+      return path + query;
+    }
```
これだとオーバーロードの意味が無いので、昨日やってた方向で行こうと決めたのに・・・なぜこう変えた・・・ :weary:

:innocent:

sd-tomita replied:
> これだとオーバーロードの意味が無いので、昨日やってた方向で行こうと決めたのに

すみません、記憶にないです。
↓私はこの認識でしたが聞き漏らしていたっぽいです。
https://github.com/SunriseDigital/cs-sdx/pull/16#issuecomment-138511917

こちらも修正します。

sd-tomita committed:
変数名にスネークケース使っていたところを修正

とりあえずひととおり置換しました。

sd-tomita commented:
1e6e6d2 以外の修正作業は別途 fix ブランチ切って対応しようとおもいます。

sd-tomita committed:
BuildQueryString() 内で引数の型で処理を分岐していたのをやめる

以前の動作に戻しています。
BuildQueryString() は、あくまで「引数を元にURL文字列を生成する」動作だけを
担当することになるので、メソッドの責任としてもこのほうがよさそうです。

sd-tomita committed:
Dictionary の連結処理を LINQ を使ったコードに変更

sd-tomita committed:
URL文字列の組み立てを StringBuilder() を使った方法に変更

String クラスは文字列操作のたびに新しいオブジェクトが生成されるらしい。
StringBuilder クラスの場合はオブジェクトの内容を変更するだけで
新しいオブジェクトは作らないようなので、こちらに変更しました。

sd-tomita committed:
無駄な Dictionary を作っている箇所があったので削除

ついでにコメントも何か変だったので修正

sd-tomita committed:
this.paraData に何もオブジェクト入ってなかったので追加

完全に凡ミス

sd-tomita committed:
foreach ステートメントを List.ForEach で書き換えました

中の処理で特に戻り値が無いので今回は Select 等の出番は無さそうと思い
普通に ForEach を使いました。

sd-tomita committed:
引数が String[] 時の処理を追加

List用の処理から単に引数の型を変えるだけでいけた。

sd-tomita committed:
Concat でキーが重複した場合の挙動を確認するテストを追加

結果例外だったので修正が必要そうです。

sd-tomita committed:
Concat でエラーになるため、事前に重複キーを持つパラメータを除外

sd-tomita committed:
変数の整理

・冗長な変数名を修正
・不要な変数を削除

sd-tomita committed:
必要そうなプロパティを追加

とりあえず Visual Studio 先生の機能で生成だけしてみました。
エラーになるのでテストは一旦全てコメントアウトしています。

sd-tomita committed:
簡単なプロパティを追加

機能を使ってみる目的で、簡単そうな Host, Path の2つを
まずは追加してみました

sd-tomita committed:
パラメータ用のプロパティを追加

sd-tomita committed:
不要な Get,Set メソッドを削除

プロパティに移行したため削除しました

sd-tomita committed:
通信プロトコル用データ追加

sd-tomita committed:
Scheme の取得テストを追加

sd-tomita committed:
Scheme の取得方法を変更

Protocol だと意味が異なってしまうので
Scheme で取得できるように変更しました

sd-tomita committed:
テスト内容の変更

"Protocol" → "Scheme" に変更するにあたってテスト内容を変更。
なお、列挙体とプロパティの名前がかぶってしまうので、プロパティについては
名前を Name と Type に分けました。

Sdx.Web.Url の中に作らず、Sdx.Web.Scheme にすればかぶらないですが
Url クラス内の Scheme にしないと、http,https 以外のスキームもないと
意味がおかしくなるので、とりあえずこのようにしています。

sd-tomita committed:
とりあえずテストが通るようにURLスキーム周りの処理を実装

Url.SchemeType は Sdx.Web.Url.Scheme 自体が返ってくるような
イメージがするので名前が微妙かも。
Sdx.Web.Url.Scheme.Http --> "http" という取得ができたほうが
使い勝手は良さそう。ただそれならDictionaryで済むような気がするので
かえって微妙かもしれない。

ひとまずテストが通ったので一旦コミット

sd-tomita committed:
スキーム管理方法変更

列挙子は使わないことになったのでコードを修正

sd-tomita committed:
System.Uri はコンストラクター以外では使わないようにしました

各プロパティが read-only で、もしあとからスキームだけを変更したりしたい場合に
対応できなかったので、メンバ変数に保存しておくようにしました。

sd-tomita committed:
不要な using の削除

sd-tomita committed:
クエリの無いURLをコンストラクタの引数に渡した際のテストを追加

後から順次セットしていく使い方もありうるのでテスト

sd-tomita committed:
テストをもう1パターン追加

こういうケースもあるかもしれない

sd-tomita committed:
パラメータが単数、なし の場合にエラーになっていたのを修正

sd-tomita commented:
# Sdx.Web を Sdx プロジェクト内へ移動する

Sdx.Web.Url の修正作業が一段落したら作業する(TODOにも追記済み)
ライブラリを相互参照できるようにすることが目的。
結果、cs-sdx ソリューション内にはプロジェクトが Sdx と UnitTest の2つだけになるようにする。

ソリューションエクスプローラー上での移動と、それだけでは元ファイルは移動しないので、元ファイルを手で移動させるのも必要。(たぶん。削除したときがそんな動作だったので、移動もおそらくおなじ)

sd-tomita committed:
Merge branch 'fix/ApiAccessManagementLibrary/fixUrlLibrary' into proj/ApiAccessManagementLibrary

sd-tomita committed:
Merge branch 'master' into proj/ApiAccessManagementLibrary

Conflicts:
	UnitTest/UnitTest.csproj

sd-tomita committed:
テスト用メソッド名が雛形のままになっていた箇所を修正

すみません漏れてました。

sd-tomita committed:
プロパティを自動実装に書き換えました

http://ufcpp.net/study/csharp/oo_property.html#auto
すごい、メンバ変数が全部いらなくなった。。。

sd-tomita committed:
パラメータの追加、削除を行うテストを追加

sd-tomita committed:
空のDictionary、値が空のDictionaryをセットしてみるテストを追加

念のため 空文字 と null の2パターン用意

sd-tomita committed:
存在しないキーを削除しようとするテストを追加

何もしない という動作を想定。
指定されたキーは存在しないことをエラーなどで返すことも考えましたが、
パラメータの内容はライブラリの使用者がセットしたものだけが全てなので
今回はわざわざ存在しないことをエラーで通知する必要は無いかと考えました。

sd-tomita committed:
存在しない値を取得するテストを追加

現状は例外になる。

sd-tomita committed:
存在しないキーを指定した場合のテストは一旦中止

存在しないキーを指定した場合、例外になるほうが正しい挙動に思えたため一旦コメントアウト。
ここで null が返ってくると、もともと値が null のパラメータと区別がつかなくなる。

sd-tomita committed:
期待した例外が出るかどうかのテストを追加

sd-tomita committed:
配列、連想配列のクエリ文字列を使ったテストメソッドを追加

テスト動作自体はこれから足します

sd-tomita committed:
Sdx.Web.Url.Param を private なプロパティにするための修正

とりあえず従来の Dictionary<string,string> のまま、Getメソッド、Setメソッドに
各種を差し替えるところまで。

sd-tomita committed:
配列パラメータ用のテストケース追加

sd-tomita committed:
連想配列用のテストを追加

C#ではクエリ文字列で [] をキーに添えても配列が表現できるわけではないので
連想配列でデータを送信する場面がそもそもないかも。このテストはいらなそうだったら削除予定。

sd-tomita committed:
不要なテストメソッドを削除

7925fcc でいらなそうだったら削除予定と書いていますが
やはり不要になりましたｗ

sd-tomita committed:
テスト追加したので一旦コミット

まだ増えたり減ったりするかとおもいますが、一旦ここで記録を残しています

sd-tomita committed:
メソッド名をテスト内容に即した名前に変更

ここでテストする内容の本質は Array がどうこうって話ではなく、
同じキーのパラメータを複数持っている場合の挙動についてということなので
変更しました。

sd-tomita committed:
同一キーで異なる値を持つクエリに対応するための作業中

・パラメータを List<Dictionary<string,string>> で管理
・コンストラクター、Getメソッド、Setメソッドをパラメータ管理方法の変更に追従

sd-tomita committed:
テストした内容をコミットしていなかったのでコミット

とりあえず Get、Set メソッドの通常挙動のみ確認しています。

sd-tomita committed:
Buildメソッドの挙動を新しいパラメータ管理方法に追従しました

sd-tomita committed:
存在しないキーで値をGetしようとした際のテストを追加

そもそもパラメータを List<Dictionary<string,string>> で管理していることは
コードを使う側からは知る由もないので、テストがあったほうが良い

sd-tomita committed:
存在しないキーで値をGetしようとしたら例外を投げるようにしました

このままnewしただけのlistを返しても、このlistで何かしようとすれば
どの道例外が投げられるので、この段階で例外を投げています

sd-tomita committed:
First()を不要な箇所で使っていたので修正

ここはキーで決め打ちできるので変更しました。

sd-tomita committed:
コンパイルできるようになったのでテストのコメントアウトを解除

Paramの管理方法変更中はコンパイルができなくなっていたので
しばらくコメントアウトしていました。

sd-tomita committed:
同じキーの値を複数持てるようにしたことで挙動が変わるのでテスト追加

Setメソッドと、Build(Dictionary) で挙動が異なるようになる。
(今までは上書きのみだったが、Setは上書きしないようする必要ができたので
そのあたりもテストするため)

sd-tomita committed:
Build(Dictionary) は上書き動作になるよう修正

要は Build の動作は既存と同じままになります。

sd-tomita committed:
空のDictionaryをBuildの引数に渡すとテストがコケていたのを修正

sd-tomita committed:
空文字、null 絡みのテストが少し足りなかったので別途追加

sd-tomita committed:
Param の型を List<KeyValuePair<string,string>> に変更

ひとまずプロパティの型と、コンストラクターのみ修正

sd-tomita committed:
ParamListの型変更に伴い、Setメソッドを修正

また、これまでのSetParamが追加の動作のみ行っていたので
別途AddParamメソッドを追加

動作としては以下のとおり。(おいおいメソッドのドキュメントに追記する)

SetParam･･･指定されたキーが存在している場合、古いものは全て削除してから追加する
AddParam･･･追加のみ行う

sd-tomita committed:
GetParamList だと、ParamListプロパティを丸ごと返すように見えるので名前変更

もし、ParamListを丸ごと返すような動作を追加するときの名前にも困るので
今のうちに名前を変えました

sd-tomita committed:
Get系メソッドの動作を現在のParamListの型に合わせて修正

sd-tomita committed:
GetParam の動作がやや特殊なのでメソッドコメントで補足

最初の要素の値だけが返って来る という動作が
コードを使っているところだけ見てもパッとはわからないので。

sd-tomita committed:
パラメータの型やメソッドの動作変更に伴ってテストを修正

各テストに期待する結果も変わる箇所があるのでまずそこを修正。
新しいメソッドに対応したテストも追加

sd-tomita committed:
Build 系メソッドをParamListの型変更に動作を合わせました

sd-tomita committed:
パラメータ削除用メソッドを作成

sd-tomita committed:
特定のKeyに一致するValueの数を控えておく処理を追加

GetParams の List.ForEach 中に break する方法が
見つからなかったので、そこだけはとりあえず foreach ステートメントを
使用しました。

sd-tomita committed:
参照型のクラス Tuple の List でパラメータを管理するように変更

より安全と思しきため変更
構造体 KeyValuePair よりもメモリ使用量も少なく済む　はず

sd-tomita committed:
空のDictionaryが渡された場合に備えた処理を追記

実際はあまりそういう使い方はされないと思われるが、念のため追加

sd-tomita committed:
Tuple の Item1, Item2 が何かをコメントで追記

どっちがどっちだかわからんｗ

sd-tomita committed:
Merge branch 'fix/ApiAccessManagementLibrary/addTest' into proj/ApiAccessManagementLibrary

sd-tomita committed:
Merge branch 'master' into proj/ApiAccessManagementLibrary

sd-tomita committed:
QueryStringList は使う必要がなかったので削除

sd-tomita committed:
共通のインターフェース IEnumerable<string> を引数に指定

List<string>, Array で別々のメソッドに同じ実装をしていたので修正

sd-tomita committed:
複数のパラメータをセットするテストを追加

Dictionaryだと順番が保証されていないことは認識済みです。
とりあえずテストでコケた場合は、目視で期待したパラメータの追加・削除が
行われていれば一旦よしとします。

sd-tomita committed:
Build 時に複数のパラメータを処理できるように修正

sd-tomita committed:
HasParam 用のテストを追加

sd-tomita committed:
指定されたキーがパラメータ内に存在するかを返すメソッドを追加

sd-tomita committed:
Get メソッドの引数に存在しないキーを渡すテストを追加・修正

GetParam() と GetParams で期待する挙動が異なるので
それぞれのテストを追加・既存テスト修正

sd-tomita committed:
Get メソッドの引数に存在しないキーを渡した際の挙動を変更

GetParams は要素がひとつも無い List を返し、
GetParam は例外を投げる

sd-tomita committed:
微妙な変数名を修正

ループ回数の制御に使うのがメインとはいえ、RoopCount だと
限定されすぎるように思えるので、ParamCount に変更

sd-tomita committed:
コメント修正

sd-tomita committed:
色んな形式のURLを使ったテストケースを作成中

途中ですが、一旦コミットします

sd-tomita committed:
テスト内容を修正

もう少しいろいろやってみました。
ここではエンコード動作をチェックしたかったので、ユーザー情報やポートを省略しないURLについては
また別の話と思い、ひとまずこのテストからは削除しました。
(ユーザー情報やポートを変えたいという状況があるかはなんともいえないですが。)

sd-tomita committed:
エスケープが必要なURL文字列を渡された場合の動作を追加

力技ですがとりあえずテストは通るようになりました。

sd-tomita committed:
value が null の際にエラーになっていたので修正

今までのテストを実行し忘れていて気づくのがちょっと遅れました。
実際の使用では起こりえない気はするものの念のため対応しています。

sd-tomita committed:
より簡潔に書けるためnull合体演算子を使ってみました

sd-tomita committed:
HasParam 用のテストが足りなかったので追加

sd-tomita committed:
set 時にエスケープされた key の判定がうまくできていなかったので修正

System.Uri.Query プロパティ自体がエスケープされた文字列を返しているので
ParamList に格納されている各パラメータは key も value もエスケープ済みの
状態になっている。なので、ここでもエスケープ処理が必要でした。

sd-tomita committed:
古いコメントが残っていたので削除

ただ、本当はここでローカルパス等のエスケープもやっておくべきな気がする。。。

sd-tomita committed:
ローカルパスはエスケープした状態で持つように変更

たまにローカルパスが日本語のサイト見かけるので一応。
System.Uriクラスのプロパティ、メソッドではどう出力しても
クエリ以外はエスケープされていないので手動で対応。

sd-tomita committed:
Merge branch 'fix/ApiAccessManagementLibrary/fixAfterCodeReview' into proj/ApiAccessManagementLibrary

sd-tomita committed:
Sdx.Web の移動準備にあたり開発中のクラスを一旦削除

移動したらmasterにマージを行うため、Sdx.Web.Api はひとまず削除します。

sd-tomita committed:
Sdx.Web プロジェクトを Sdx へ統合

ライブラリを相互参照できるようにするための変更です

sd-tomita committed:
Merge branch 'master' into proj/ApiAccessManagementLibrary

Conflicts:
	Sdx/Sdx.csproj

sd-tomita commented:
Sdx.Web を Sdx に統合した状態を master に反映するため、このブランチは一旦閉じます。
作業再開時は新しく proj ブランチを作って対応します。
